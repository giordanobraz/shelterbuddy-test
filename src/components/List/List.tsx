import {
  Accordion,
  AccordionDetails,
  AccordionProps,
  AccordionSummary,
  AccordionSummaryProps,
  Avatar,
  Box,
  Button,
  ButtonProps,
  Typography,
  accordionClasses,
  accordionSummaryClasses,
  buttonClasses,
  styled,
} from "@mui/material";
import { useEffect, useState } from "react";

import { ChevronRight } from "@mui/icons-material";
import { Data } from "../../utils/createData.util";
import { GridExpandMoreIcon } from "@mui/x-data-grid";
import { Pagination } from "../Pagination";
import PlaceholderImg from "../../assets/placeholder.png";
import { fetchAnimalsPhotoData } from "../../services/axios";
import { getComparator } from "../../utils/comparator.util";

interface ListProps {
  filteredAnimalList: Data[];
}

const BASE_PHOTO_URL = "https://shelterbuddy-us-uat.shelterbuddy.io";

const CustomAccordion = styled(Accordion)<AccordionProps>({
  [`&.${accordionClasses.root}`]: {
    backgroundColor: "#f8f8fc",
    borderRadius: "12px",
    margin: "12px 0",
    padding: "12px",
    boxShadow: "none",
    [`&:before`]: {
      display: "none",
    },
  },
});

const CustomAccordionSumary = styled(AccordionSummary)<AccordionSummaryProps>({
  marginBottom: 0,
  [`& .${accordionSummaryClasses.expanded}`]: {
    marginBottom: "0",
  },
  [`& .${accordionSummaryClasses.expandIconWrapper}`]: {
    backgroundColor: "#002fd6",
    color: "#fff",
    borderRadius: 100,
  },
});

const CustomButton = styled(Button)<ButtonProps>({
  [`&.${buttonClasses.root}`]: {
    backgroundColor: "#002fd6",
    color: "#fff",
    fontSize: "16px",
    fontWeight: 600,
    borderRadius: "8px",
  },
});

function buildDetails(title: string, data: string) {
  return (
    <Box
      display="flex"
      flexDirection="row"
      gap={1}
      fontSize="16px"
      fontWeight={500}
      color="#707183"
    >
      <Typography>{title}: </Typography>
      <Typography
        sx={{
          color: "#000121",
        }}
      >
        {data}
      </Typography>
    </Box>
  );
}

export function List({ filteredAnimalList }: ListProps) {
  const rows = filteredAnimalList;
  const [page, setPage] = useState(0);
  const [photos, setPhotos] = useState<any>([]);

  useEffect(() => {
    fetchAnimalsPhotoData().then((response) => {
      setPhotos(response?.Data);
    });
  }, []);

  const slicedRows = rows
    .slice()
    .sort(getComparator("asc", "name"))
    .slice(page * 10, page * 10 + 10);

  return (
    <Box>
      {slicedRows.map((row, index) => {
        const photo = photos.find((photo: any) => photo.Animal.Id === row.id);

        return (
          <CustomAccordion key={index}>
            <CustomAccordionSumary
              expandIcon={<GridExpandMoreIcon />}
              aria-controls="panel1a-content"
              id="panel1a-header"
            >
              <Box
                component="div"
                display="flex"
                flexWrap="wrap"
                alignItems="center"
                gap={2}
              >
                {photo ? (
                  <Avatar
                    alt={row.name}
                    src={`${BASE_PHOTO_URL}${photo.Photo.replace(
                      "/1024---n",
                      ""
                    )}`}
                  />
                ) : (
                  <Avatar alt={row.name} src={PlaceholderImg}></Avatar>
                )}
                <Typography>{row.name !== "z" ? row.name : "-"}</Typography>
              </Box>
            </CustomAccordionSumary>
            <AccordionDetails>
              <Box display="flex" flexDirection="column" gap={1}>
                {buildDetails("Type", row.type)}
                {buildDetails("Breed", row.breed)}
                {buildDetails("Gender", row.gender)}
                {buildDetails("Color", row.color)}
                <CustomButton>
                  Details <ChevronRight />
                </CustomButton>
              </Box>
            </AccordionDetails>
          </CustomAccordion>
        );
      })}
      <Pagination
        currentPage={page}
        setCurrentPage={setPage}
        totalRegisters={rows.length}
        rowsPerPage={10}
      />
    </Box>
  );
}
