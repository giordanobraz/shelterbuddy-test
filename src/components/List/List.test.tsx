import { render, screen } from "@testing-library/react";

import { ComponentProps } from "react";
import { Data } from "../../utils/createData.util";
import { List } from "./List";
import userEvent from "@testing-library/user-event";

describe("List", () => {
  it("should render with data", () => {
    const { container } = getRenderer();
    expect(container).toMatchSnapshot();
  });

  it("should open accordion when clicked", () => {
    getRenderer();
    const accordion = screen.getByRole("button", { name: "Lion Lion" });
    expect(accordion.getAttribute("aria-expanded")).toBe("false");
    userEvent.click(accordion);
    expect(accordion.getAttribute("aria-expanded")).toBe("true");
  });

  it("should close accordion when clicked", () => {
    getRenderer();
    const accordion = screen.getByRole("button", { name: "Lion Lion" });
    userEvent.click(accordion);
    expect(accordion.getAttribute("aria-expanded")).toBe("true");
    userEvent.click(accordion);
    expect(accordion.getAttribute("aria-expanded")).toBe("false");
  });
});

const animals: Data[] = [
  {
    id: 1,
    name: "Lion",
    breed: "Poodle",
    color: "Black",
    gender: "Male",
    type: "Dog",
  },
  {
    id: 2,
    name: "Brad",
    breed: "Husky",
    color: "White",
    gender: "Male",
    type: "Dog",
  },
];

function getRenderer({
  filteredAnimalList = animals,
}: Partial<ComponentProps<typeof List>> = {}) {
  return render(<List filteredAnimalList={filteredAnimalList} />);
}
