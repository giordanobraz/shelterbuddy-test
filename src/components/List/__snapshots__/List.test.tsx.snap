// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render with data 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-7ku0nv-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-controls="panel1a-content"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-1weusj3-MuiButtonBase-root-MuiAccordionSummary-root"
        id="panel1a-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <div
            class="MuiBox-root css-jpm4l0"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
            >
              <img
                alt="Brad"
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="placeholder.png"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Brad
            </p>
          </div>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="panel1a-header"
              class="MuiAccordion-region"
              id="panel1a-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
              >
                <div
                  class="MuiBox-root css-1821gv5"
                >
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Type
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Dog
                    </p>
                  </div>
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Breed
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Husky
                    </p>
                  </div>
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Gender
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Male
                    </p>
                  </div>
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Color
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      White
                    </p>
                  </div>
                  <button
                    class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-ve7jsm-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    Details 
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ChevronRightIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-7ku0nv-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-controls="panel1a-content"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-1weusj3-MuiButtonBase-root-MuiAccordionSummary-root"
        id="panel1a-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <div
            class="MuiBox-root css-jpm4l0"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
            >
              <img
                alt="Lion"
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="placeholder.png"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Lion
            </p>
          </div>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="panel1a-header"
              class="MuiAccordion-region"
              id="panel1a-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
              >
                <div
                  class="MuiBox-root css-1821gv5"
                >
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Type
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Dog
                    </p>
                  </div>
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Breed
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Poodle
                    </p>
                  </div>
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Gender
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Male
                    </p>
                  </div>
                  <div
                    class="MuiBox-root css-1r2odvj"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Color
                      : 
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-16m92ws-MuiTypography-root"
                    >
                      Black
                    </p>
                  </div>
                  <button
                    class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-ve7jsm-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    Details 
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ChevronRightIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1u96uvg-MuiStack-root"
    >
      <button
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        1
      </button>
    </div>
  </div>
</div>
`;
